### ISSUE TITLE
Add functionality to iterate over a StarRecord

### ISSUE THREAD

**Issue by csbrown** [BelgianBiodiversityPlatform/python-dwca-reader#99]
## Description:

I need to be able to iterate over Core records JOINed on Extension records.  My specific current use case is to iterate over a Taxon core JOINed with a VernacularName extension.  Similar functionality exists in the [`dwca-io.StarRecordIterator`](https://github.com/gbif/dwca-io/blob/master/src/main/java/org/gbif/dwc/StarRecordIterator.java).

It looks like this should be easy enough to implement, utilizing the hash index functionality in `files.py`.

I will be taking up this task, as I need it for my own project.  I can put all of the functionality into a new file emulating the structure of `dwca-io.StarRecordIterator`.  I will try to conform to existing code styling.   

I am happy to contribute back to upstream.  I don't see any Contributing.md anywhere, so any advice on adhering to any informal contribution guidelines for this project is appreciated.

## Deliverables:

1) A class that enables a simple JOIN on the coreid between arbitrary files in a DWCA archive.
    - implements the python iterable interface.
    - implements the python context manager interface
2) An automated test for the JOIN functionality on an existing appropriate test archive (e.g. dwca-2extensions.zip)

**Comment by csbrown** [comment]
Advice on a recommended development workflow?

Something like:

1) Fork this repo
2) clone the fork
3) iterate locally on development of my fork
    - ???? Should I repeatedly run setup.py to get my new changes?
    - ???? Should I do this in a venv?
    - ???? How do I run the tests?
4) push my branch to my fork
5) PR my fork to y'all.

**Comment by niconoe** [comment]
Thanks @csbrown, I merged this! Sorry for not answering your question about the development workflow, I was quite busy and only noticed it now. It seems you figured it out yourself in the meantime.

FYI, there's some development workflow information [here](https://python-dwca-reader.readthedocs.io/en/latest/contributing.html), but this should probably be improved.

**Comment by csbrown** [comment]
No worries about the wait, it's easy enough to swap out my fork and the official repo in our `requirements.txt`  :)


I noticed that `nose` is having some errors since Python 3.10 and doesn't seem to be actively maintained anymore.  I ran the tests with `pytest` from the root folder and it "just worked".

Hopefully I can contribute more.  Not sure if we'll end up needing things like write functionality.

Cheers,
-Scott!
